### ISSUE TITLE
Instantiating Multiple Providers with a loop

### ISSUE THREAD

**Issue by jaken551** [hashicorp/terraform#19932]
### Current Terraform Version

```
Terraform v0.11.11
```

### Use-cases

In my current situation, I am using the AWS provider so I will scope this feature request to that specific provider, although this may extend to other providers as well.

I am attempting to create resources in multiple AWS accounts.  The number of the accounts ranges from 0....x and will be dynamic. I would like to be able to instantiate multiple providers which can assume a role in each account, and in turn, create resources with the associated provider without hard-coding providers for each subsequent account.

For example, something like this:
```
variable "accounts" {
  type    = "list"

  default = ["123456789012", "210987654321"]
}
variable "names" {
  type = "map"

  default = {
    "123456789012" = "foo"
    "210987654321" = "bar"
  }
}
provider "aws" {
  count   = "${length(var.accounts)}"
  alias   = "${lookup(var.names, element(var.accounts, count.index))}"
  
  assume_role {
    role_arn = "arn:aws:iam::${element(var.accounts, count.index)}:role/ASSUMEDROLE"
  }

resource "aws_instance" "web" {
  count         = "${length(var.accounts)}"
  provider      = "aws.${lookup(var.names, element(var.accounts, count.index))}"
  ami           = "${data.aws_ami.ubuntu.id}"
  instance_type = "t2.micro"

  tags = {
    Name = "HelloWorld"
  }
}

**Comment by debovema** [comment]
With a more generic formulation, it would be great to have a ```count``` attribute in **providers**, not only to instantiate multiple times the same provider but also to choose between several **providers**: some having a zero count, some having a non-zero count.

**Comment by hhh0505** [comment]
Has anyone ever found any kind of workaround for this kind of scenario? We have multiple AWS accounts and would like to be able to deploy modules across all accounts without having to have repeatable code for each account.

**Comment by sc250024** [comment]
@jaken551 Have you tried this with Terraform 0.12 ? Did you make any progress? I ask because I'm looking for the same functionality.

**Comment by kuwas** [comment]
We have the same issue, basically the ask here is to allow `provider` blocks to be dependent on computed values from other resources (such as creating a new AWS account then using the account ID to assume role) and also support `for_each` and `count` for providers.

**Comment by autodeck** [comment]
@jaken551 @kuwas - I'd also like to see provider blocks support for_each. My use case being creating aws_ses_domain_identity resources in multiple regions. These regions being defined by a list variable i.e.
```
resource "aws_ses_domain_identity" "example" {
  for_each = var.ses_regions

  provider = each.value
  domain = var.ses_domain
}

**Comment by shankarsundaram** [comment]
One workaround is, you need to  maintain separate directory for each account and call the same terraform module by passing correct values to the modules / keep some default values. to run plan  / apply in all directory use wrapper such as terragrunt.
https://blog.gruntwork.io/terragrunt-how-to-keep-your-terraform-code-dry-and-maintainable-f61ae06959d8
https://davidbegin.github.io/terragrunt/use_cases/execute-terraform-commands-on-multiple-modules-at-once.html

**Comment by rgevaert** [comment]
I'm an enterprise user and doing a for_each over the vault provider would reduce a lot of code if you could just do:

```
variable "namespaces" {
  type        = set(string)
  default     = []
  description = "Names to be created"
}

resource "vault_namespace" "namespace" {
  for_each = var.namespaces
  path     = each.key
}

provider "vault" {
  for_each = var.namespaces
  alias = each.key
  namespace = each.key
}
```

**Comment by Roxyrob** [comment]
Terraform deployment environment protection using separated Cloud Providers accounts  is a great method to reduce configuration/deployment error Blast Radius. E.g. different AWS accounts can be used to make deployment environments of any kind (devel, stage, prod, sandbox), with potentially a lot of account with same "**common configuration**".

Without a solution like for_each in provider block lot of identical code must be replicated, raising  a lot the risk of typing errors on addition of each new account/environment.
Example:
 - using aws profile in provider alias to assume role in accounts
 - a master account manage e.g. identity account and any other single deployment/environment account (devel, prod, sandboxN, ecc.)
 - creating user and permissions in identity account
 - creating base common config for each account (assumable roles with policies, terraform state with S3/DynamoDB, etc.)

Based on account info like:
```
locals {
  accounts = {
    "devel" = { ... }
    "prod" = { ... }    
    "sandbox1" = { ... }    
    "sandbox2" = { ... }    
    ...
  }
}
```
for_each in provider allow to achieve a common configuration using DRY code blocks, looping on accounts as base key for many decoupled resources:
 ```
  module "example1" {
     source = "..."
     for_each local.account
     provider {
       aws = "aws.${each.key}"
     }
     ... create assumable roles or terraform states (S3/dunamoDB) or other common resources in the same way making DRY code ...
  }
```

**Comment by bryankaraffa** [comment]
+1 on this.  We would like to use Terraform to deploy/manage IAM Roles in a list of AWS Accounts -- assuming a known role in each account for access.  Below is an example implementation that would work for us with Terraform >v0.13 and the existing `module.for_each` -- just also need an equivalent `provider.for_each`:

```terraform
## Just some data... a list(map())
locals {
  aws_accounts = [
    { "aws_account_id": "123456789012", "foo_value": "foo",    "bar_value": "bar"    },
    { "aws_account_id": "987654321098", "foo_value": "foofoo", "bar_value": "barbar" },
    ]
}
## Here's the proposed magic... `provider.for_each`
provider "aws" {
  for_each = local.aws_accounts
  alias = each.value.aws_account_id
  assume_role {
    role_arn    = "arn:aws:iam::${each.value.aws_account_id}:role/TerraformAccessRole"
  }
}
## Modules reference the provider dynamically using `each.value.aws_account_id`
module "foo" {
  source = "./foo"
  for_each = local.aws_accounts
  providers = {
    aws = "aws.${each.value.aws_account_id}"
  }
  foo = each.value.foo_value
}
module "bar" {
  source = "./bar"
  for_each = local.aws_accounts
  providers = {
    aws = "aws.${each.value.aws_account_id}"
  }
  bar = each.value.bar_value
}
```

**Comment by DonBower** [comment]
Allowing alias in the provider config would also make for cleaner code:

```
variable awsEnvironments {
  type = list
  default = [
    "paas-lusw2",
    "gi-luse1",
    "gi-lusw2"
  ]
}

variable awsRegion {
  type = map
  default = {
    paas-luse1 = "us-east-1"
    paas-lusw2 = "us-west-2"
    gi-luse1   = "us-east-1"
    gi-lusw2   = "us-west-2"
  }
}

provider "aws" {
  profile                             = "paas-luse1"
  alias                               = "paas-luse1"
  region                              = var.awsRegion[var.awsEnvironments["paas-luse1"]]

  ignore_tags {
    key_prefixes                      = ["Core:"]
  }
}

provider "aws" {
  count                               = length(var.awsEnvironments)
  profile                             = var.awsEnvironments[count.index]
  alias                               = var.awsEnvironments[count.index]
  region                              = var.awsRegion[var.awsEnvironments[count.index]]

  ignore_tags {
    key_prefixes                      = ["Core:"]
  }
}

module "peering" {
  source                              = "../tf-source/vpc/peering"
  for_each = var.project

  providers = {
    aws.local = aws.each.value.localEnvironment
    aws.peer = aws.each.value.peerEnvironment
  }

  localEnvironment                    = each.value.localEnvironment
  peerEnvironment                     = each.value.peerEnvironment

  localSuffix                         = each.value.localVPCSuffix
  peerSuffix                          = each.value.peerVPCSuffix

}

``` 

In this case, the module `"../tf-source/vpc/peering"` would be expecting two AWS Providers, **local** and **peer**.

**Comment by ajbouh** [comment]
@DonBower AFAIK, alias *is* allowed in the providers config for modules

**Comment by DonBower** [comment]
> @DonBower AFAIK, alias _is_ allowed in the providers config for modules

It is.  but not as a variable....
```
Error: Variables not allowed

  on provider.tf line 27, in provider "aws":
  27:   alias                               = var.awsEnvironments[count.index]

Variables may not be used here.
```

**Comment by kuwas** [comment]
@ajbouh @DonBower
Yes, you can pass aliased providers into modules, but the providers and their aliases are still statically defined. Furthermore, when using `for_each` with modules, you must pass in the **same** provider instance to every module instance within the loop.

I recall reading a comment from @apparentlymart on the reasoning around this current limitation.

Example 1: this does not work
```
provider "azurerm" {
  for_each = toset(["one", "two"])
  alias       = each.value
}
module "test" {
  for_each   = toset(["one", "two"])
  providers = { azurerm = azurerm[each.value] }
}
```

Example 2: this also does not work
```
provider "azurerm" { alias = "one" }
provider "azurerm" { alias = "two" }
module "test" {
  for_each   = toset(["one", "two"])
  providers = { azurerm = azurerm[each.value] }
}
```

Example 3: this does work
```
provider "azurerm" { alias = "one" }
provider "azurerm" { alias = "two" }
module "test" {
  for_each   = toset(["one", "two"])
  providers = { azurerm = azurerm.one }
}
```

**Comment by oliverlucasbca** [comment]
I've literally tried to this exact same thing and have been unable.

`module "peering" {
  for_each = var.vnet_peering_config
  source   = "./config/network_peering"
  providers = {
    azurerm.custom = "azurerm.${each.value.provider_alias}"
  }

  resource_group_name = module.resource_group.rg.name
  ne_virtual_network  = azurerm_virtual_network.default
  we_virtual_network  = azurerm_virtual_network.default

  remote_virtual_network_name = each.virtual_network_name
  remote_resource_group_name  = each.resource_group_name
}`

I'm actually pretty upset that I cant do it, as with peering each "peer" would need to go into another subscription!

Sigh.

**Comment by efernandes-dev-ops** [comment]
Any updates if this is on the roadmap?

**Comment by TulyOpt** [comment]
Hi,

Passing a provider as variable would be very useful, is it doable ?

**Comment by augustgerro** [comment]
Any updates? Also depends from this ... 
For AWS provider the only way to use multi-regions for me is dublicate code many times

```tf
provider "aws" {
	  alias = "euwestprovider"
	  region = "eu-west-1"
	  ...
}
	
	provider "aws" {
	  alias = "eucentralprovider"
	  region = "eu-central-1"
	  ...
	}
  
	resource "aws_instance" "ec2-eu-west"{
	  provider = "aws.euwestprovider"
	  ami = "ami-030dbca661d402413"
	  instance_type = "t2.nano"
	}
	
	resource "aws_instance" "ec2-eu-central" {
	  provider = "aws.eucentralprovider"
	  ami = "ami-0ebe657bc328d4e82"
	  instance_type = "t2.nano"
	}

```

**Comment by Lroca88** [comment]
Any updates from the team? 
This thread has been opened since 2019, please give it some attention.

**Comment by ced3eals** [comment]
+1 on the topic.
Can you give us some feedback? We are using multiple accounts in my company, and it will be great to be able to loop through organization accounts to create the same resources. for_each in providers block is really needed.
Thanks.

**Comment by dhdersch** [comment]
In my humble but perhaps irrelevant opinion, Terraform 1.0 shouldn't be released until this feature is implemented.

**Comment by egee-irl** [comment]
Hey all 👋 - I found a solution that works perfectly to create multiple sets of S3 buckets (or any resource) across multiple regions using multiple [providers](https://www.terraform.io/docs/language/meta-arguments/module-providers.html) and [modules](https://www.terraform.io/docs/language/modules/index.html). It's not as clean as provider interpolation would be, but it **does** work.

An example using two S3 buckets in _different_ regions:

```
provider "aws" {
  alias   = "use1"
  region  = "us-east-1"
}

provider "aws" {
  alias   = "use2"
  region  = "us-east-2"
}

module "s3_use1" {
  source    = "./s3s"
  providers = {
    aws.region = aws.use1
  }
  bucket_name = "my-bucket-in-use1"
}

module "s3_use2" {
  source    = "./s3s"
  providers = {
    aws.region = aws.use2
  }
  bucket_name = "my-bucket-in-use2"
}

```
_in ./s3s:_

```
provider "aws" {
  alias = "region"
}

variable "bucket_name" {
}

resource "aws_s3_bucket" "my_s3_bucket" {
  bucket   = var.bucket_name
  provider = aws.region
}
```
This solution was inspired by [this](https://stackoverflow.com/questions/48632797/using-terraform-to-manage-multiple-aws-regions) Stack Overflow answer.

**Comment by NeverOddOrEven** [comment]
With the `azurerm` module, it is not possible to create a resource group in a specific subscription unless you make the provider explicit. This limitation makes it impossible to iterate over a map or list of configuration data, when the underlying subscription is also parameterized.

```terraform
provider "azurerm" {
  features {}
  alias           = "specific_subscription"
  subscription_id = "00000000-0000-0000-0000-000000000000"
}

resource "azurerm_resource_group" "foo" {
  name     = "foo-rg"
  location = "centralus"

  provider = azurerm.specific_subscription
}
```

Would love suggestions on how to achieve this, or why this approach is flawed according to Terraform design philosophies.

**Comment by yongzhang** [comment]
I have the same issue, I can't find a way to checkout workspaces from different resource groups.

**Comment by ziggythehamster** [comment]
I know this sounds a bit crazy, but I solved this using M4 macros. You can use `esyscmd` to do whatever you want, and then the output of that will be handled as macros. So, for instance, here's a simplified `main.tf.m4` that does something for every AWS region:


```m4
/**
 * ******** DO NOT EDIT THIS FILE ********
 *
 * This file is auto-generated. If a new region has launched, please run make.
 */
dnl
dnl You can edit this file, though. The end of the file will shell out to the AWS CLI and
dnl magically build main.tf based on the current regions available.
define(`PROVIDER', `provider "aws" {
  version = "= 3.42.0"

  # picks up most of its configuration from the environment
  alias  = "some-special-identifier-prefix-$1"
  region = "$1"
}
')dnl
dnl
dnl The macro which drops all of the stuff for each region.
define(`REGION', `
#
# BEGIN AUTO-GENERATED RESOURCES FOR $1
#

PROVIDER($1)
# Put macros here for the resources you want to create

#
# END AUTO-GENERATED RESOURCES FOR $1
#
')dnl
dnl
dnl This is the magical part where we get all of the regions and output macros for each
esyscmd(`aws ec2 describe-regions --region us-east-1 --query "Regions[].RegionName" | jq -r ".[] | \"R\" + \"EGION(\" + . + \")\""')dnl
```

And a Makefile:

```make
# This Makefile is used to build main.tf, which sets up XXX.
# Rerun make whenever new regions are made available.

# Ensure required binaries are present
REQUIRED_BINS := aws jq m4
$(foreach bin,$(REQUIRED_BINS),\
	$(if $(shell command -v $(bin) 2> /dev/null),,$(error Please install `$(bin)`)))

# If you just type make, build this
MAIN = all

# Build main.tf
all: main.tf
.PHONY: all

# Clean the built file
clean:
	rm -f main.tf
.PHONY: clean

# The rule that builds the main.tf file
%.tf: %.tf.m4
	m4 $@.m4 > $@
```

You could expand on this to have e.g., a HCL or JSON file which lists out info for different accounts and then builds appropriate providers and dependent module/resource instances. The downside is this can't be dynamic, since you have to re-run make to update the Terraform file.

**Comment by PCjrBasic** [comment]
I dealt with it by using a shell script that calls terraform to output a map variable that contains the account structure, it then iterates over each element (account) to build a terraform environment for that environment (copy in the base code, copy in the variable structure, build a backend file so everyone has their own S3 state, etc.) and then launch a new terraform instance for every single one, capture the output and exit codes, and then dump the output from each one back to the caller and determine it's own exit code based on the exit code of all the instances it launched (if anyone exited with an error, exit with an error...). It's a waste - terraform to call terraform because terraform can do a loop in it's provider block.

**Comment by yves-vogl** [comment]
I'd like to add another use cases to justify this demand.

1.
When you setup an AWS Organization and follow best practices, you often want to create a centralized network account and share the VPCs via AWS RAM. You also might want to place a private route53 zone in each account you share the VPC to (e.g. apps.integration.private, database.integration.private) and associate the VPCs with the zones. 
Currently you need to configure a provider for each target account and pass it to a module. Neither the account creation nor the „passing to the module“ can be dynamic so you end up with a lot of duplicated code.

2.
When working with AWS organization accounts you might want to bootstrap each account after creating it the same run (e.g. setting IAM account alias, creating some more restrictive IAM provisioner roles than the OrganizationAccountAccessRole) - this is currently not possible because you need to retrieve account IDs and setup a provider for each account with OrganizationAccessRole.


Today I'll give this merged PR a try (https://github.com/hashicorp/terraform-provider-aws/issues/14932) - this could potentially save some of those issues.

**Comment by mcandre** [comment]
What's the point of making a DSL when you can't loop properly? Might as well have stuck with plain YAML.

**Comment by Nuru** [comment]
@apparentlymart @jbardin Can we have an update on this? Other issues have been closed in favor of using this one for tracking, but I cannot find a progress update in the past year on this issue.

Last I heard this was [planned for Terraform 1.0](https://github.com/hashicorp/terraform/issues/24476#issuecomment-700368878) (or maybe delayed until is more accurate), which of course you know is now out. We need it to generically work with an arbitrary set of accounts.

**Comment by jbardin** [comment]
Hi @Nuru,

Sorry, I don't have any update at the moment. While it seems like a simple request on the surface, the underlying architecture of terraform is not suited for using providers in this manner. What this essentially means is that it's going to be a very large project which must complete in resources with many other large projects. While this is definitely a desired feature, and fulfilling this use case has a high priority, I cannot say which release may be targeted for implementation.

**Comment by Nuru** [comment]
@jbardin Thank you for the update. 

### I have a proposal.

It seems to me that Hashicorp is thinking about implementing "provider for_each" in a way largely if not completely consistent with how "resource for_each" is implemented, and that is "very hard". So I suggest you look at alternatives. 

Here is one alternative. I can give you others. The main point I am making is to have the team think outside the box.

If you read this and other issues, you see that the workaround people are using is code generation. Also, people are willing to accept significant restrictions on the feature, such as that all parameters be specified as input variables (since that is effectively the restriction they have on code generation). 

So how about instead of `for_each` you come up with some other term, like `init_each`. You can even create a new keyword such as `provider_template` to replace `provider`. The way it would work is that `provider_template` would use values available at ***init*** time to generate a set of providers during `terraform init`. Variables not available at init time would be copied as variables, and variables available at init time would be expanded. You could require that the input to `init_each` be a set of strings, and that the keys will be the provider `alias`. Init-time values can be set via `provider.auto.tfvars` or all the other ways Terraform allows values to be set.

I believe this would cover more than 80% of the use cases without being "a very large project".  You certainly could release it as experimental and see what the feedback is like, and use the experimental status to de-prioritize support for edge cases. 

<details><summary>Example code:</summary>

```hcl

locals {
  regions = toset(["us-east-2", "us-west-2"])
}

provider_template "aws" {
  init_each = local.regions
  region    = each.value

  role_arn = var.role_arn
}
```
would be equivalent to

```hcl
provider "aws" {
  region = "us-east-2"
  alias  = "us-east-2"

  role_arn = var.role_arn
}

provider "aws" {
  region = "us-west-2"
  alias  = "us-west-2"

  role_arn = var.role_arn
}
```

</details>

### Extra credit

To make this even more valuable and useful, I would ask you to introduce a new provider attribute, `enabled`, that would behave like this:
- The provider must exist and be downloaded at init time, regardless of the setting. 
- The `enabled` value must be derivable before the provider needs to be initialized.
- When `enabled` is true (the default), provider works as it does now.
- When `enabled` is false, at the time the provider would be initialized, initialization is skipped and the provider is removed from state as if it were not included in any file. 

This is a feature/attribute that can be provided by core Terraform and applied to any provider without the provider author needing to modify anything. It is important here because it would allow a common configuration that essentially says "create every possible provider" and then allow the ones not needed to be disabled at plan time. (This feature would also be useful on its own, for the cases where providers need credentials but the user does not have credentials and does not need the features that provider supplies. I have hit that several times where the provider author does not allow for that case and throws an error because of missing credentials.)

**Comment by nikolay** [comment]
@jbardin Well, maybe it needs a lot of resources, but the sooner it is done, the better. At least allow some pre-processing with Terraform Cloud as we pay tons of money for it, but cannot run a preprocessor, unlike people who use Terraform for free. We simply cannot have dynamic Kubernetes clusters as each cluster needs a hardcoded provider! At least fix your Kubernetes provider if you're not going to fix the provider implementation! At the end of the day, most people use Kubernetes today and many who use it would want to have dynamic clusters and not hardcoded ones.

**Comment by AndreiPashkin** [comment]
I don't see why Terraform can't have all constructs including providers to be dynamic. The only restrictions for any of TF constructs are that they should be able to execute without erros and yield their part of the plan. Why must there be anything static? That makes no sense.

**Comment by faGH** [comment]
Why was this closed? Is this supported now? I tried a mere day ago with no joy :(

**Comment by rgevaert** [comment]
> Why was this closed? Is this supported now? I tried a mere day ago with no joy :(

it's not closed, you are referring to https://github.com/hashicorp/terraform/issues/29500, which mentioned this ticket.

**Comment by whoaks** [comment]
Any update on dynamic provider in azure?

**Comment by chaddjohnson** [comment]
Really need this. It's annoying to have to do this:

```
module "api_us_east_1" {
  source = "./region"
  providers = {
    aws.region = aws.us-east-1
  }
}

module "api_eu_west_1" {
  source = "./region"
  providers = {
    aws.region = aws.us-east-1
  }
}

module "api_ap_northeast_1" {
  source = "./region"
  providers = {
    aws.region = aws.ap-northeast-1
  }
}
```

**Comment by nikolay** [comment]
If providers are so hard to change, why don't they introduce the concept of a connection, which could be built from the ground up and be dynamic? At the end of the day, we don't need dynamic providers - we need dynamic connections and constructs for us to pass them to resources.

**Comment by nikolay** [comment]
To clarify, a connection belongs and is specific to a provider.

**Comment by dogfish182** [comment]
I would also love to see this feature. something I was trying when I ran into this was attempting to build a global network of transit gateway peers. 
Transit gateway peering (and other aws features) have an 'requester/accepter' way of working, so cross region peering requires 2 providers. I was trying this NON working code when I ran into this issue. 

```
# This cant work due to
# https://github.com/hashicorp/terraform/issues/19932
# providers cannot be iterated.
module "transit_gateway_peers" {
  source = "./modules/transit_gateway_peering"
  for_each = local.global_peers_map
  providers = {
    aws.requester = each.requester
    aws.accepter = each.accepter
  }
  peer_region = each.accepter_region
  peer_transit_gateway_id = each.peer_transit_gateway_id
  transit_gateway_id = aws_ec2_transit_gateway.transit_gateway_ew1.id
}

locals {
  global_peers = [
    {
      requester = "aws.eu-west-1"
      accepter = "aws.eu-central-1"
      accepter_region = "eu-central-1"
      requester_region = "eu-west-1",
      peer_transit_gateway_id = "the-foreign-id"
      transit_gateway_id = "the-local-id"
    }
  ]
  global_peers_map = {for peer in local.global_peers : format("%s-%s", peer.requester_region, peer.accepter_region) => peer}
}
```
I got very confused because in older versions terraform requires you define the provider alias with a string, in later versions that's gone. 

in earlier versions the for_each command just bombs if provider block is included. 

my 'ignorant of the underlying workings' wish would be that I could just use string formatting to define the provider and iterate over it as above. Then I could make a clever local to figure out every combination of peering required and deploy it all with 1 module. 

Now i need to calculate how many peers there will be and continue making new module calls for each peer as the global network grows.

**Comment by ajbouh** [comment]
My needs are centered around making "multiregion" modules.

I initially was writing these by hand but this was tedious and error prone.

So I've written a tool that takes an implicitly single provider module and does some code generation to emit a multiprovider module that references it. The generated module has configurable variables and outputs.

The contract is fairly straightforward.

It takes as input:
- module source
- list of provider aliases
- list of "global" variables names
- list of "per provider" variable names
- list of "map aggregated" output names
- list of "concat aggregated" output names

And generates as output a module with:
- "global" variables that are "passed through" to per provider module
- "per provider" variables that are maps from provider alias name to a value specific to that provider
- "map aggregated" outputs are maps keyed by provider alias with corresponding values from the inner module output of the same name
- "concat aggregated" outputs are lists made by concat of all the inner module outputs together

**Comment by CMorton737** [comment]
I feel like I wind up on this page one way or another once a month..  So many use-cases, so little movement..

**Comment by edwdev** [comment]
Managing a deploying resources across an AWS organisation, would greatly benefit from this feature. One use case here, is being able to iterate over all accounts and VPC's to create a list of all VPC CIDR's for IP management.

**Comment by nikolay** [comment]
@CMorton737 They have [160 open engineering positions](https://www.hashicorp.com/jobs/engineering) - NY being so severely understaffed, I don't doubt that we will have to wait years, not months for this!

**Comment by brunoscota** [comment]
I really need this. PLEASE, make this possible.

**Comment by patresso-sb** [comment]
Doing my part in this request, I would really appreciate this feature. Creating resources and associating them across multiple accounts in one apply is a c/p nightmare at this point. 
As was said a bit above, every resource that have requester/accepter parts can't be done the right way.

**Comment by nikolay** [comment]
At least, for those who made the mistake to sign an annual contract with HashiCorp for TFC - add support for Terragrunt like [env0](https://www.env0.com/)! If you guys don't want to support Terragrunt, at least support your own baby [cdktf](https://www.terraform.io/cdktf) in TFC so that we work around the copypasta needed with providers! Oh, you don't support anything that's beta (unlike pretty much any other technology company out there)! Many APIs we use with Kubernetes are even labeled `alpha`, for example! It's been months waiting for even variable sets in TFC to be exposed in the TFE provider!

**Comment by nikolay** [comment]
@patresso-sb This alone turns IaC into IaCP (infrastructure-as-copypasta)!

**Comment by atanas-ampeco** [comment]
We are really struggling without this feature for the same reasons mentioned in the comments above. All sorts of workaround and ugly fixes must be made just to avoid the copy-paste of code. Please fix this

**Comment by eredi93** [comment]
hitting this issue as well 😞

**Comment by benvbr** [comment]
Hitting this limitation as well. Any news on this?

**Comment by ecout** [comment]
Is there a technical reason why this hasn't been fixed in over two years? For our use case all we're asking is for terraform to support a functionality similar to Cloud Formation Stack Sets without having to recur to that resource, why is that so difficult?

**Comment by bryankaraffa** [comment]
Reminder all, if we want to increase priority of this issue please help and click the 👍 on the initial post: https://github.com/hashicorp/terraform/issues/19932#issue-396653063

**Comment by sc250024** [comment]
> Is there a technical reason why this hasn't been fixed in over two years? For our use case all we're asking is for terraform to support a functionality similar to Cloud Formation Stack Sets without having to recur to that resource, why is that so difficult?

I remember seeing (and don't quote me on this; just trying to remember that one brief moment a while ago) that it has to do with how the Terraform model works under the hood. They (the developers) make an assumption that any `provider` block must be loaded first before any of the HCL language constructs (like `for_each`, `count`, etc.) are processed and rendered. I think that's sort of why it's taken this long. To do this issue is definitely not trivial, otherwise it would have been solved a while ago.

**Comment by ecout** [comment]
> > Is there a technical reason why this hasn't been fixed in over two years? For our use case all we're asking is for terraform to support a functionality similar to Cloud Formation Stack Sets without having to recur to that resource, why is that so difficult?
> 
> I remember seeing (and don't quote me on this; just trying to remember that one brief moment a while ago) that it has to do with how the Terraform model works under the hood. They (the developers) make an assumption that any `provider` block must be loaded first before any of the HCL language constructs (like `for_each`, `count`, etc.) are processed and rendered. I think that's sort of why it's taken this long. To do this issue is definitely not trivial, otherwise it would have been solved a while ago.

It makes sense considering the Stateful nature of Terraform. Thanks for the reminder.

**Comment by ecout** [comment]
> Reminder all, if we want to increase priority of this issue please help and click the 👍 on the initial post: [#19932 (comment)](https://github.com/hashicorp/terraform/issues/19932#issue-396653063)

Done! And thank you for your code example!

**Comment by gtmtech** [comment]
Cross-posting from #31069 as requested

I work on an enormous hybrid cloud platform, - it consists of 3 clouds, hundreds of accounts/projects, multiple environments, regions and so on... Some simple terraform installations would not have a need for a solution in this space, but if you are a B2B provider, or an internal service provider to a very large company (e.g. 50,000 employees+), you are likely to hit the difficulty with how to structure all of your terraform in a way that scales well.

Terraform best practice dictates that a single terraform run (and thus state file) should manage a moderate, but not excessive number of resources - perhaps 100 is a reasonable amount of terraform resources per state file. At some point as you scale out, you have to make decisions on how to split resources across multiple terraform state files.

In a large multi-account AWS setup for example, you might make a reasonable decision to start splitting state across account boundaries, especially if a lot of AWS accounts are very similar and have similar resources set up (cloudtrail, networks, configs, iam etc.)

This also seems reasonable because in order to terraform across 2 accounts in AWS, you need 2 providers. A provider is intimately tied to a set of credentials and a context with which to access the AWS API, and those set of credentials or that context stipulates the target AWS account you are terraforming in.

Providers support part-parameterisation, so you can inject things like the role_arn, or this or that credential and you have parameterised code that can be run against each of your AWS accounts, and this results in a statefile per account. This is the typical way I have seen companies use terraform through my client engagements.

As that scales up though, you get further problems. Instead of 2 accounts, suppose you are terraforming 1000. That means 1000 terraform runs, and that means you need an automated orchestrator to keep them all in sync, as its too big a task for a human - step in things like terragrunt, or some enterprise offerings to help out.

However whilst those sort the running of lots of terraform runs, another problem starts to creep in. Developers tend to work on features, not on accounts, and terraform itself may not run completely as desired across all 1000 accounts due to engineer fault, network timeouts, aws api rejections, race conditions. Some AWS resources tend to take a very long time to instantiate and change - such as AWS Microsoft Active Directory at around 50 minutes. During such a change, all 1000 accounts are potentially "locked out" from other feature development. In a large team of engineers with a lot of features to change, waiting for hours for terraform runs to finish is not great.

So you might think to split out certain products from other products and create separate runs for them so they dont lock out the entire state for others whilst updating. Now instead of 1000 statefiles, you have 2000, and then 4000 and so on.

For each split, you may well need to introduce dependencies in your orchestrator to make sure that X happens before Y - e.g. the IAM permissions are set up before the resources that need them - or the networks are set up before the loadbalancers.

Pretty soon, the orchestrator dependencies is also creating a huge workflow that takes hours to resolve as well - So if my CI system has to run 8000 terraform runs, and there is an orchestrator dependency chain, even with some very well provisioned CI servers, I may be waiting hours for a run. In a large team and estate, this starts to cripple your productivity.

I was thinking about how best to solve some of these issues, when I realised that the main reason why the design choices had led down this path, was because there was no real way within terraform to intelligently and dynamically operate across a large number of AWS accounts. And the reason for that is because there is no support for dynamic providers.

Whilst providers can reference attributes, the number of providers is always fixed in the terraform configuration, and this means that if you want to operate on 1000 accounts in a single terraform run, you will need 1000 provider blocks. Or if you want to operate on each account using 2 IAM roles, you will need 2000 provider blocks.

Managing these provider blocks could be done by simply hardcoding them to each aws-account-id or each role-id - and maintaining a lot of these in your repository. For example, you could have a provider.account1.tf for each account, in there specify 2 provider blocks with hardcoded values for each of your roles you want to use in the respective account.

But these providers need passing through to the modules in the correct way, so you also need a module block that is going to call a module with a different set of providers for each account as well. And as with every module, you need to pass through all the variables needed too.

This works, its just a lot of files and blocks to maintain. It does however allow a different model at operating at scale, which is to now get terraform to do work across all accounts in a single terraform run, and that can be aligned to a specific feature. For example you could have a "cloudtrail" module which sets up cloudtrail in every account (in the days before the cloudtrail AWS Organizations feature as an example). You could imagine that you want these 1000 cloudtrails stored in a bucket too in a different account, so your single terraform run/statefile contains the code to create the cloudtrail bucket in some audit account with associated policies, and then create all the cloudtrails across all accounts, feeding them into the bucket, all as one feature.

This would seem pretty neat! Now with statefiles split along feature lines, because developers tend to naturally work on features and not on accounts, this aligns the developer iteration with the code iteration. Different features can be worked on in isolation (just as different developers work on them in isolation).

Onboarding new accounts and offboarding old accounts is a little trickier, as every feature needs running to onboard/offboard them in the respective account, but use of -target since each account's resources is a "module", is easy, so that works too.

It's a bit ugly though to maintain 1000s of provider blocks and associated module blocks, and the interactions between them (the dependencies that may be required between modules, like in the case above between the module doing the account which contains the cloudtrail bucket, and the module doing the accounts which contain the cloudtrail).

It would be a very nice feature to be able to iterate through a list of accounts, and generate provider blocks and associated module blocks , which means provider{} and module{} should both be made to support for_each and all the supporting ecosystem. Currently only module{} has been made to support it.

I managed to get such a system up and running using terragrunt doing the dynamic generation of provider blocks and module blocks - I put a POC here if anyone is interested: https://github.com/gtmtechltd/terragrunt-poc - it works and I was surprised at how much I loved being able to terraform a multi-account feature with different aspects in different accounts in one terraform run.

However I turn up my nose a little at such metaprogramming - programs to write programs - it's all just a bit... hard to read. I'd much rather this was supported directly in terraform. Also terragrunt doesn't really have support for datasources itself, and the generation phase has to come before any terraform runs happen, so you can't do something even cooler, which would be to query the aws_organizations object as a datasource, get all the accounts, and then just create all the dynamic providers from that, and apply your stuff everywhere.

Appreciate 99% of your userbase are probably just terraforming a small infra in dev/stage/prod setup - and we do have all sorts of terraform enterprisey things going on, but I thought it would be good to write up as a general feature request the reasons why I think dynamic providers would really help out at scale and offer some genuinely great alternatives to organising workloads in ways which are small and isolated and aligned with developer workflow.

If the terraform team have any other ideas and best practices about how to divide resources along statefile boundaries which lend themselves well to operating a large cloud estate with a large team at scale, I'd be really interested to hear from experience in the field.

Attempted Solutions
See https://github.com/gtmtechltd/terragrunt-poc for a workaround using terragrunt

Proposal
Allow provider blocks to support foreach, and be able to be dynamically made according to other datastructures (which could be just vanilla variables, rather than derived from datasources, but the icing on the cake would be if they could be made from the output of datasources too)

**Comment by MrDrMcCoy** [comment]
My use case for this would be for deploying a non-fixed number of Kubernetes clusters that you then need to deploy common manifests into (security, monitoring, ingress, etc). Each cluster created by Terraform would require its own K8S or Helm provider definition, which is extraordinarily unruly and verbose in vanilla Terraform.

**Comment by mcandre** [comment]
I've had to wrap the ungainly HCL code inside of Jinja2 as a workaround.
And explain the need for Jinia2 to several mystified developers.

Seems that a mistake in parsing logic has uniquely broken the for loop
here, compared to most other programming languages. Maybe just deprecate
that old for syntax and create a new token with a more useful loading
priority.

**Comment by nikolay** [comment]
I gave up on this! Honestly, CDK for Terraform seems to be the only (kind of) clean and HashiCorp-backed solution right now excluding Terraform Cloud support - I can generate dynamic providers and even name resources dynamically as a plus. You can have the CDKTF code as simply glue code of the HCL code organized in modules. In addition, you can break your monolithic HCL codebase into stack within the same workspace. The only drawback is that you can't really put pieces of your code in traditional Go modules and reuse them across projects.

**Comment by thunder-spb** [comment]
Oh, yeah. Suffering with the same issue. In my case, I have ECR and Replication set up. But AWS ECR Replication does not sync Repository permissions nor Lifecycle policy from the source repository to the replicated ones. And I have to precreate repository with all required permissions and lifecycles in the target regions/accounts. That also allow me to cleanup the replicated repostiories if the source repository has been deleted.... Replication regions' list could be dynamic...

 Right now, I have to make a lot of kludges to address this... :(

So, I would love to have dynamically defined providers! +1 for this change!!

**Comment by nimblenitin** [comment]
is there any workaround available to use variable in provider something like below. If yes please let me know how to do it. Currently terraform is not allowing to use variable in provider name for module. I want to create role in each member account user provides; that is my goal here. User will provide the alias which I will use to create role.

```
providers = {
    aws = "${lookup(var.names, element(var.accounts, count.index))}"
  }
```

**Comment by tmccombs** [comment]
As far as i know the only workaround is to use another tool to dynamically generate terraform config.

**Comment by jcollado** [comment]
I'm using [doit](https://pydoit.org/) and [jinja2 templates](https://jinja.palletsprojects.com/en/3.1.x/) to generate terraform files. Just in case that is useful to anybody, I've uploaded the code to [this gist](https://gist.github.com/jcollado/3caf195f616076eaf1009280e5cc307d).

**Comment by devopsinfoltd** [comment]
I know there must be lots of new feature request, but can someone from the members/contributors please update us, so we can start investing time/money to write our own dirty wrappers. Please really in need of this feature. Thanks.

**Comment by prmarino1m** [comment]
is there any movement on this it was buried so deep in the list that i couldn't find it and created a new feature request for the same thing and even included an example  where i had the for_each loop iterating on the names provided by

data "aws_regions" "all" {
  all_regions = true
  filter {
    name = "opt-in-status"
    values = ["opted-in" , "opt-in-not-required"]
  }
}

**Comment by ejtbrown** [comment]
Bumping for visibility. This issue is a significant flaw in Terraform.

**Comment by air3ijai** [comment]
It is a long post with a lot of workarounds but all of them are about external tools or services
1. [esyscmd](#issuecomment-856191323) - template engine
2. [Terragrunt](#issuecomment-1131431007) - wrapper for Terraform
3. [Jinja2](#issuecomment-1132241395) - template engine
4. [CDKTF](#issuecomment-1132360703) - Cloud Development Kit for Terraform
5. [env0](#issuecomment-1040792670) - IaC workflows
6. [gomplate](https://gomplate.ca/) - template engine

But I have another question about how to handle dynamically resources for different provider using template engine or this request when it will be implemented.
```terraform
provider "aws" {
  region = "eu-west-1"
  alias  = "eu-west-1"
}

provider "aws" {
  region = "eu-west-2"
  alias  = "eu-west-2"
}

resource "aws_acm_certificate" "primary" {
  domain_name       = "domain.com"
  validation_method = "DNS"
  provider          = aws.eu-west-2
}

resource "aws_acm_certificate" "secondary" {
  domain_name       = "domain.com"
  validation_method = "DNS"
  provider          = aws.eu-west-1
}
```

This static code works well but just for creating new resources. If we remove code for `eu-west-2`, apply will fail with the error about missing provider:
>
>╷
│ Error: Provider configuration not present
│
│ To work with aws_acm_certificate.primary its original provider configuration at provider["registry.terraform.io/hashicorp/aws"].eu-west-2 is required, but it has been removed. This occurs when a
│ provider configuration is removed while objects created by that provider still exist in the state. Re-add the provider configuration to destroy aws_acm_certificate.primary, after which you can remove
│ the provider configuration again.
╷

So, using a king of dynamic regions (when it will be implemented) will fail if provider will be removed before resource will be destroyed if we change the list of regions:
```terraform
locals {
  regions = ["eu-west-1", "eu-west-2"]
}

provider "aws" {
  for_each = local.regions
  alias   = each.value
}

resource "aws_acm_certificate" "regions" {
  domain_name       = "domain.com"
  validation_method = "DNS"

  for_each = local.regions
  provider = "aws.${each.value}"
}
```

Maybe @Nuru's [proposal](#issuecomment-877438031) with addition attribute `enabled` may solve this.

**Comment by hypervtechnics** [comment]
Maybe this can be handled similar with the [`prevent_destroy`](https://developer.hashicorp.com/terraform/language/meta-arguments/lifecycle#prevent_destroy) which has to be first set to `false`, be applied and then can actually be removed.

**Comment by prmarino1m** [comment]
Obviously there has been no movement 3 year old issue which I suspect has related older issue if you dug into all the closed issues. I'm wondering if some one from Hashicorp can give us a reason why this is hard thing to implement and maybe point out what would need to change in the code so we can contribute the required fixes to get this working. This issue has become a chain of here is how I've worked around it and people repeating the different ways they would like it to work but no one seems to be pointing out the underlying reason why it doesn't work that way already.

**Comment by prmarino1m** [comment]
by the way I'm assuming the root of the issue is lack of key information about providers in the state storage.

**Comment by Waldonutz** [comment]
+1 for this would be useful, Im trying to deploy resources to 30+ accounts atm, without blowing out my code, a simple looping mechanism like for_each for providers would be appreciated.

**Comment by kemalizing** [comment]
Any updates on this issue?

I also want to link another issue, which is directly related: https://github.com/hashicorp/terraform/issues/24476 

@teamterraform should take a look at these issues as they are one of the top 👍 issues and impacting a lot of people.

**Comment by ebarrere** [comment]
Yes please. It is very difficult to follow AWS best-practices for Organizations (i.e. create an account for everything) without being able to loop over providers.

**Comment by kehindeakala** [comment]
Any updates on this implementation please?

**Comment by pexa-akansal** [comment]
Any Updates on this?

**Comment by crw** [comment]
Thank you for your comments. We are aware of this issue. There are no updates to report at this time. 

As a reminder, please avoid "+1" and "any updates" comments, and to use the upvote mechanism (click or add the 👍 emoji to the original post) to indicate your support for this issue. Thanks again for the feedback!

**Comment by hholland-gl** [comment]
Any updates around this

**Comment by vierkean** [comment]
Hello,
I finally took the time to update from terraform version 0.11 to the latest version.
Now I have to realize that this is not so easy because in version 0.11 we pass provider alias as variable.

With which version will this be possible again?

**Comment by freakinhippie** [comment]
> Hello, I finally took the time to update from terraform version 0.11 to the latest version. Now I have to realize that this is not so easy because in version 0.11 we pass provider alias as variable.
> 
> With which version will this be possible again?

@vierkean this should probably be opened as question on the hashicorp developer forums. However, providers within modules can be defined using the `configuration_aliases` directive in the `required_providers` block. See [the docs here](https://developer.hashicorp.com/terraform/language/modules/develop/providers#provider-aliases-within-modules).

There is a notable exception to the legacy passing providers by alias which is that now all providers defined in the `configuration_aliases` must be explicitly passed into the module calls rather than inherited implicitly.

**Comment by deepbrook** [comment]
For anyone looking for a workaround, there's really only one that made sense to me:

Multiple invocations of the same plan with different inputs for each provider.

I run my configuration in a CI/CD job matrix, each with different provider credentials/configs for the terraform command's environment.

IMO, completely feasible for most use-cases. And I think in cases where it isn't, there's probably a code smell somewhere which could be taken care of with a little refactoring.

**Comment by kahawai-sre** [comment]
@deepbrook I'm likely missing something here ... but does that mean you are calling a module with the same name and provider alias, but with different provider configuration (account, subscriptionId, region etc, depending on the provider type) for each plan/apply? Does that not cause issues?
Also, if there are approvals etc for each plan, does that not become a burden as the number of pipeline runs (alias configurations) scales?

**Comment by Mike-Nahmias** [comment]
@kahawai-sre I don't use a job matrix like @deepbrook mentioned but I basically took the same approach. Yes, for me I have a separate deployment/state per provider instance (different regions). We haven't had any issues so far as we scale. For handling resource dependencies I use Terragrunt. I use it for other stuff too but it makes it really easy to handle the inter-deployment dependency piece.

**Comment by deepbrook** [comment]
> @deepbrook I'm likely missing something here ... but does that mean you are maintaining a single deployment (plan/apply/state file) per distinct provider instance in your environment? If that is the case, and if there are approvals etc for each plan, does that not become a burden as the number of pipeline runs (alias configurations) scales? Also, does that approach not make it harder to orchestrate resource dependencies across multiple target deployment scopes i.e. through a single deployment?

That's exactly what that means, yes :)

My deployments do not run at a big scale, to be fair. As for approvals, I do not have those - I deploy infrastructure to preprod and run tests to ensure all features required are reachable/configured correctly, and then deploy them.

I do have some dependencies between the deployments; since my infrastructure isn't managed in a single repository, but several ones (basic platform infra like networking and IAM roles in one, k8s cluster and database in another one each, etc) It's only matter of triggering down-stream pipelines (I like pipelines :P ). 

@Mike-Nahmias approach using terragrunt sounds like a good alternative if you're a mono repo kinda person. :)

**Comment by dbhaigh** [comment]
Same issue - New day...    I have a load of terraform deploying resources to several environments, each requiring their own provider alias

I want to dedupe this lovely collection of n00dles to a single resource block with a for_each loop that iterates over a local or var that provides the required provider alias for each resource

I don't want to have to use modules for this as I feel it's an ugly way of doing this and will make things far harder and more complicated than the how I'm doing it now (separating each resource block, and any associated resources, out into it's own file), which makes it very easy to deploy new resources of the same type (ctrl-C, Ctrl-V - change the names of the guilty, and et viola! there you are)

What I want is to be able to have something like this

resource "resource_type" "resource_name" {
    for_each = local/var.resource_values
    provider = each.provider
    ...
}

Without being told that "each" is an invalid provider reference

This is a major hole, and an old hole - one that should've been backfilled in years ago

Can we please get some action on this?  and save me from having to go down the ugly bunny hole of modules?

TIA

**Comment by duxing** [comment]
coming back to this issue after a couple years and still not supported :(

**Comment by nikolay** [comment]
I guess, we have no other option but to move this issue to https://github.com/opentofu/opentofu - maybe they can implement this must-have feature now that they have a larger staff dedicated to Terraform!

**Comment by raelix** [comment]
I don't think here there will be updates soon...I decided to move on Pulumi which is not affected by this problem.

**Comment by nikolay** [comment]
The strategy is intentionally limiting Terraform and solving these limitations by design only in TFC and patenting them!

**Comment by Poltergeisen** [comment]
700 upvotes on this makes me think that it's something we should at least get some sort've update / explanation on why this isn't possible. 

A ton of the Terraform configurations just "work" with this type of syntax, why not providers as well?

**Comment by nikolay** [comment]
@Poltergeisen They want you to use their upcoming Terraform Cloud Stacks feature, so they have no incentive to fix this for all of us!

**Comment by nikolay** [comment]
The new pricing model of Terraform Cloud is crazy! They bill you per managed resource! So, now we have to move pieces to alternatives as we want to avoid paying per label we manage in GitHub with the GitHub provider! Pay for each Terraform Workspace, run triggers, policies, and other features we manage with Terraform and not via ClickOps! **That! Is! Ridiculous!** It seems like we all should focus on [OpenTofu](https://opentofu.org/) and open-source alternatives of Terraform Cloud/Enterprise such as [Terrakube](https://terrakube.org/), [Digger](https://www.digger.dev/), [Hatchet](https://hatchet.run/), or [OTF](https://docs.otf.ninja/)! Enough _[abuse and negligence]_ is enough! I wonder why they wasted efforts to no-code Terraform when they can't make IaC Terraform work smoothly!

**Comment by nikolay** [comment]
![2023-11-15_16-30](https://github.com/hashicorp/terraform/assets/2882/7c75bd60-d580-4f0d-9df2-10a81866a855)

**Comment by nikolay** [comment]
Thank you, @hashicorp!

**Comment by nitoxys** [comment]
> 700 upvotes on this makes me think that it's something we should at least get some sort've update / explanation on why this isn't possible.
> 
> A ton of the Terraform configurations just "work" with this type of syntax, why not providers as well?

There's another request like this, but the developer stated that providers are loaded on init and it would have to be rearchitected. Technically provider_override could be an option within a provider.

**Comment by osterman** [comment]
Fwiw, we've solved this problem using [`atmos`](https://atmos.tools) with Terraform and regularly deploy Terraform root modules (what we call components) to dozens or more of accounts in every single AWS region (e.g. for compliance baselines), without HCL code generation. IMO, even if this requested feature existed in terraform, we would _not_ use it because it tightly couples the terraform state to multiple regions, which not only breaks the design principle for DR (that regions share nothing), but makes the blast radius of any change massive and encourages terralythic root module design. In our design pattern, we instantiate a root module once per region, ensuring that each instantiation is decoupled from the other. The one exception to this is when we set up things like transit gateways (with hubs and spokes), then we declare two providers, so we can configure source and target destinations. This ensures no two gateways are tightly coupled to each other. 

**TL;DR:** I acknowledge why at first glance not supporting multiple providers in a loop seems like an awful limitation and why some things would be simplified if it were supported; that said, we deploy enormous infrastructures at [Cloud Posse](https://github.com/cloudposse) in strict enterprise environments, and don't feel any impact of this inherent limitation.

**Comment by yves-vogl** [comment]
I could not agree more to @osterman 

I'm following this thread since a long time and I also came here because I thought that I needed this feature. Meanwhile I think that this feature could possibly be a sign of a bad design. As @osterman said - just think of the enourmous blast radius.

One could argue that there are certain cases where you e.g. want to dynamically create provider instances e.g. in an AWS organization to assume the OrganizationAccountAccessRole for bootstrapping things. But once you think about managing the whole lifecycle of hundreds of those accounts you'll realize that this does not scale. Instead I use AFT with AWS Control Tower to customize those accounts. 

And maybe a few words on all the hate against Hashicorp. I also think that their pricing model
is insane and the times I had contact with the sales team (and even afterwards with their legal team) I can tell you that they somehow
overestimate what they bring to the table as their are valuable alternatives nowadays. But please also keep in mind that there are lot of companies which make thousands of dollars each day without giving anything back to the community. 
Everyone who demands feature should ask himself how much he gave back in form
of contributions of donation before starting rude demands and spreading hate.

Yes, Hashicorp is bad a communication. And in my opinion they somehow now try to squeeze a lot of money out of their customers. But please do not forget what their impact on out community had been. I remember the old days back in around 2010 when Mitchel created Vagrant. It was the first time we could manage those VM in a way which did not suck. 

Having said all the those things… I understand the demand and I feel that the lack of communication from Hashicorp contributed to this situation. 

They better could have pointed out the reasons this feature is not available and demonstrate how to solve this problem by proper design.

Love matters,
Yves

**Comment by yves-vogl** [comment]
By the way… to understand what's going on in FSF and what's the „Free Rider“ problem check the great posts from @adamhjk on X:

https://x.com/adamhjk/status/1726311265604374750?s=46

He wrote three parts.

**Comment by SamuelMolling** [comment]
The forum is quite large, do we have any way to iterate the providers of a resource? Has anything like this been made available yet?

**Comment by air3ijai** [comment]
One of the [comment with a recap](#issuecomment-1287311175).

**Comment by dbhaigh** [comment]
This post - https://github.com/hashicorp/terraform/issues/19932#issuecomment-1815164297  only three posts back in the thread, and on this page, answers your question, had you actually bothered to read it

**Comment by afrazkhan** [comment]
There's another use case which I think hasn't been mentioned: Optional resource creation within a module.

I have a module which mostly provisions AWS resources, but there's a single Github resource that is created conditionally (if a particular key is present in one of the module's variable objects). It's likely that many of the people using the module aren't using Github, or don't need this particular feature / resource, so I don't want to add it to the required providers block.

**Comment by aleksasiriski** [comment]
I've also encountered this limition trying to achieve multi-region deploys for my app, I've ended up using TF templates with Terragrunt to generate Terraform code simulating the desired for_each behavior. It's ugly and doesn't provide any IDE niceness a normal .tf file would, but it [works](https://github.com/hearchco/infra/blob/main/live/prod/agent/terragrunt.hcl#L105)

**Comment by KeksBombe** [comment]
Please fix...

**Comment by mcandre** [comment]
I don't understand why the industry has chosen to adopt a tool with a broken for loop. And when you try to explain why you're wrapping Terraform in jinja2, engineers get mad. I'd just as soon use Ansible... or a shell script.

**Comment by aleksasiriski** [comment]
Take a look at OpenTofu, I believe they're adding support for this in the upcoming release.

**Comment by rgeraskin** [comment]
@mcandre terramate is a great workaround for this limitation instead of jinja2

**Comment by air3ijai** [comment]
> @aleksasiriski can you provide some link to opentofu docs about that ?

Probably this one - https://github.com/opentofu/opentofu/issues/300, and they also have a [backlink](https://github.com/opentofu/opentofu/issues/300#issuecomment-1776983419) to this issue, right after [this comment](https://github.com/hashicorp/terraform/issues/19932#issuecomment-1667278260).

**Comment by amirasyraf** [comment]
It seems Hashicorp decided to implement this feature as part of stacks: https://developer.hashicorp.com/terraform/language/stacks/use-cases#deploy-across-regions

... which currently is only available on their HCP Terraform service, with no guarantee this feature will come to the Community edition.

In other news, OpenTofu is officially going to support this natively: https://github.com/opentofu/opentofu/releases/tag/v1.9.0-alpha1

**Comment by scottt732** [comment]
OpenTofu 1.9 was released a few days ago. https://github.com/opentofu/opentofu/releases/tag/v1.9.0

**Comment by air3ijai** [comment]
**Terraform**
- Started: `Jan 7, 2019`
- Closed: `?`
- Comments: `109 comments`
- Released: `?`

**OpenTofu**
- Started: [`Aug 31, 2023`](https://github.com/opentofu/opentofu/issues/300#issue-1883864383)
- Closed: [`Dec 3, 2024`](https://github.com/opentofu/opentofu/issues/300#event-15514843086)
- Comments: `51 comments`
- Released: [`Jan 09, 2025`](https://github.com/opentofu/opentofu/releases/tag/v1.9.0)

**Comment by zoezhangmattr** [comment]
will terraform consider adding this feature. 👁

**Comment by NateLangloisProscia** [comment]
So, if OpenTofu did this, why can't native terraform do it in the same way?

To any detractors of this request who are suggesting that the myriad of use cases people brought aren't real/valid enough; the syntax already exists to support this (multiple statically defined aliased providers), it's just an inferior syntax for many of these very real+valid use cases. It's "working" fine enough for their use cases today, but the syntax limitation here is the main thing holding it back from being a good experience.

This issue has almost 1000 upvotes. In my opinion, if native terraform is going to survive, the trend of adding more tight coupling and API complexity via Terraform HCP exclusive patterns for these kinds of things needs to stop, and instead be focusing on native terraform missing syntax. I am not trying to trivialize the work needed to implement this, but after 6 years of what seems like neglect for this request based on the history in this issue, along comes OpenTofu and delivers it during its existing of a little over 16 months. It's making it harder to justify continuing to pay for Terraform Cloud/HCP and using native terraform.

If there are or are not plans to replicate with native terraform what OpenTofu did, can this answer please be made more visible in this issue? If yes, any information to help inform how soon we might see this would also be great to see. If no, please explain why. Thanks!

**Comment by ricardbejarano** [comment]
We do [this](<https://github.com/cisco-open/stacks/blob/main/docs/3.2.2.%20Terraform%20provider%20generation.md>) with Stacks. YMMV.

**Comment by Poltergeisen** [comment]
> We do [this](https://github.com/cisco-open/stacks/blob/main/docs/3.2.2.%20Terraform%20provider%20generation.md) with Stacks. YMMV.

This isn't what we're asking for. I don't want to pay for HCP to be able to use a basic feature that tofu already added

**Comment by kitos9112** [comment]
> We do [this](<https://github.com/cisco-open/stacks/blob/main/docs/3.2.2.%20Terraform%20provider%20generation.md>) with Stacks. YMMV.

There are lots of ways to tackle this feature request outside Terraform. One could argue that's not a good practice and use other tools instead (e.g. Terragrunt, custom Terraform code generator...), but the point is to have a native terraform solution that leverage the existing HCL domain-specific language we all are familiar in vanilla Terraform.

**Comment by Poltergeisen** [comment]
I don't want to change my entire deployment stack and release pipelines to
integrate terragrunt just because I want a single feature that was only
rolled out for the cloud

**Comment by ricardbejarano** [comment]
Yeah I get, I'm just sharing an alternative, this thread's been open for 6 years, might as well start looking into these workarounds instead. 🤷‍♂️

**Comment by zoezhangmattr** [comment]
any updates pls 👀

**Comment by air3ijai** [comment]
For AWS that could be solved by the [[Major Version]: Terraform AWS Provider v6.0.0 #41101](https://github.com/hashicorp/terraform-provider-aws/issues/41101)?
> [[Enhancement]: Add region argument/parameter where appropriate #27758 (comment)](https://github.com/hashicorp/terraform-provider-aws/issues/27758#issuecomment-2587102363)

**Comment by ricardbejarano** [comment]
Correct

**Comment by akloss-cibo** [comment]
My tagline should be  *"been using bash-generated .tf files since 2023... "*

**Comment by jdeus** [comment]
This is required to be able to dynamically target different Kubernetes clusters using the kubectl and kubernetes providers

**Comment by tymik** [comment]
> > For AWS that could be solved by the [[Major Version]: Terraform AWS Provider v6.0.0 #41101](https://github.com/hashicorp/terraform-provider-aws/issues/41101)?
> 
> From my understanding this would solve the cases that concern multi-region, but the problem for multi-account remains the same.

yup.

And for me there's an edge case, where I want to merge 2 very similar modules into single one, but one of them is for cross-account scenario and uses providers.
currently there's no way to do it in a non-breaking manner (so all currently defined resources using the module need no code changes), as I would need to use a conditional provider meta argument in resource's definition, which is not possible in any way.

Luckily the scale I have to support is low enough, so I can bring the definitions to the required configuration, but if the module was heavily utilised, then doing this and supporting all the teams to be aware of what changes, why and how to work with the new change, would be a challenge.

**Comment by es-cs1** [comment]
Hey All,

I see its 2026 and no movement on this?

I see OpenTofu has done the work
https://opentofu.org/blog/opentofu-1-9-0/

<img width="1499" height="222" alt="Image" src="https://github.com/user-attachments/assets/5219b3a3-6e78-42ab-8a6a-83797fdec19e" />

Are we able to backport in this functionality?

**Comment by david-garcia-garcia** [comment]
This is probably never coming to Terraform because it is part of Hashicorp Stacks commercial offering: 

https://developer.hashicorp.com/terraform/language/stacks

I have to say that migrating to OpenTofu has been easy and straightforward. Just a sensation, but looks like things are smoother. And if you already use terragrunt throughout your infra, it's as easy as adding this to the hcl file: 

```
terraform_binary = "tofu"
```

This can aid in easily having a mixture of Opentofu and Terraform.

**Comment by punmechanic** [comment]
Definitely a choice to mark commentary as spam. @david-garcia-garcia's comment is not spam and a suggestion on "this is how you can do this today" seems appropriate given this issue has been open for 7 years.

Our use-case is that we want to be able to set annotations on kubernetes service accounts a general cloud security repository that has elevated access to those kubernetes clusters. We should not need to maintain a list of every EKS cluster at every account just to be able to do this. Tofu works for us.
